### PR TITLE
[SPARK-9634][SPARK-9323][SQL] cleanup unnecessary Aliases in LogicalPlan at the end of analysis

### PR DESCRIPTION
**Pull request by cloud-fan**
Also alias the ExtractValue instead of wrapping it with UnresolvedAlias when resolve attribute in LogicalPlan, as this alias will be trimmed if it's unnecessary.

This also fixes https://issues.apache.org/jira/browse/SPARK-9323 as `DataFrame.resolve` won't return unresolved expression now.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @JoshRosen , this anwsers your last question at https://issues.apache.org/jira/browse/SPARK-9323
:)

**Comment by cloud-fan** [comment]
Seems master is broken?

**Comment by cloud-fan** [comment]
retest this please.

**Comment by marmbrus** [comment]
I think this works, but is there a reason to not just strip all `UnresolvedAlias`es (and possibly nested `Alias`es) from the plan like we do with subqueries right after analysis?

**Comment by cloud-fan** [comment]
Hi @marmbrus , we strip `UnresolvedAlias` from aggregation and projection at [Analyzer](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L140), which works well for SQL. It will be good to make this rule more general but I'm afaid that will be hard to write. So I think fixing the different behavior between DataFrame and SQL is also reasonable, i.e. making `DataFrame.resolve` return resolved expressions.

**Comment by marmbrus** [comment]
Why not this after analysis?
```scala
plan transformAllExpressions {
  case UnresolvedAlias(child) => child
  case Alias(child, name) => Alias(child transform { case Alias(c, _) => c }, name)
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -955,6 +957,18 @@ class Analyzer(
         Project(p.output, newPlan.withNewChildren(newChild :: Nil))
     }
   }
+
+  object CleanAlias extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions  {
+      case UnresolvedAlias(child) => child match {
+        case g: GetStructField => Alias(g, g.field.name)()
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Is this duplicated with `ResolveAliases`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -81,6 +81,8 @@ class Analyzer(
       UnresolvedHavingClauseAttributes ::
       HiveTypeCoercion.typeCoercionRules ++
       extendedResolutionRules : _*),
+    Batch("Cleanup", fixedPoint,
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I guess I was thinking we would do this in the same place that we remove subqueries.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -955,6 +957,18 @@ class Analyzer(
         Project(p.output, newPlan.withNewChildren(newChild :: Nil))
     }
   }
+
+  object CleanAlias extends Rule[LogicalPlan] {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: There are possibly other rules here (`ResolveSortOrders`) that have unnecessary logic if we are going to add this rule.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Maybe just search for `UnresolvedAlias` in this file and make sure we are consistent and only handle this in one place.

**Comment by marmbrus** [comment]
Thanks for continuing to work on this :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We remove subqueries in optimizer, but `UnresolvedAlias` is another story, we have to strip(resolve) it before `CheckAnalysis`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Sorry I made a mistake here, we should just remove the `UnresolvedAlias`, just like what we did in `Analyzer.ResolveReferences`(we call `trimUnresolvedAlias` there).

The key problem is we have 2 code path to resolve `UnresolvedAttribute`, one is `Analyzer.ResolveReferences`(last case) and the other is `DataFrame.resolve`, so we need to make them consistent, i.e. call `trimUnresolvedAlias` at `DataFrame.resolve` or abstract this logic for sql and dataframe.

**Comment by cloud-fan** [comment]
cc @marmbrus , I think this solution is better :)

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 4]
```diff
@@ -119,4 +119,19 @@ class AnalysisSuite extends AnalysisTest {
...
+    var plan = testRelation.select(((a + 1).as("a+1") + 2).as("col"))
+    var expected = testRelation.select((a + 1 + 2).as("col"))
+    checkAnalysis(plan, expected)
+
+    plan = testRelation.groupBy(a.as("a1").as("a2"))((min(a).as("min_a") + 1).as("col"))
+    expected = testRelation.groupBy(a)((min(a) + 1).as("col"))
+    checkAnalysis(plan, expected)
+
+    // CreateStruct is a special case that we should not trim Alias for it.
+    plan = testRelation.select(CreateStruct(Seq(a, (a + 1).as("a+1"))).as("col"))
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: test `CreateStructUnsafe` too?

**Comment by marmbrus** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 5]
```diff
@@ -1135,17 +1135,18 @@ class DataFrame private[sql](
    * @group dfops
    * @since 1.3.0
    */
-  def withColumn(colName: String, col: Column): DataFrame = {
+  def withColumn(colName: String, col: Column, metadata: Option[Metadata] = None): DataFrame = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: cc @rxin , in MLlib sometimes we need to set metadata for the new column, thus we will alias the new column with metadata before call `withColumn` and in `withColumn` we alias this clolumn again. Here I added a new parameter to allow user set metadata.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Can you do this in a different PR?  Also we should do it without using `Option` and default arguments so that it works well in Java.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -968,3 +961,61 @@ object EliminateSubQueries extends Rule[LogicalPlan] {
     case Subquery(_, child) => child
   }
 }
+
+/**
+ * Cleans up unnecessary Aliases inside the plan. Basically we only need Alias as a top level
+ * expression in Project(project list) or Aggregate(aggregate expressions) or
+ * Window(window expressions).
+ */
+object CleanupAliases extends Rule[LogicalPlan] {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Can you remove everything except for this and the related tests?  I'd like to pull this into the release branch without new features.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I have opened https://github.com/apache/spark/pull/8159 to improve the `withColumn`, but left the code here to see if we can pass the tests.

This PR did 2 things:

1. use `Alias` instead of `UnresolvedAlias` when resolve nested column in `LogicalPlan.resolve`
2. clean unnecessary aliases at the end of analysis

If we only do `1`, some tests will fail as we need to trim aliases in the middle of getField chain. If we only do `2`, it can't fix any bugs. So I put them together here.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I've open #8215 which is basically your patch without the mllib changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -968,3 +961,61 @@ object EliminateSubQueries extends Rule[LogicalPlan] {
...
+      Aggregate(grouping.map(trimAliases), cleanedAggs, child)
+
+    case w @ Window(projectList, windowExprs, partitionSpec, orderSpec, child) =>
+      val cleanedWindowExprs =
+        windowExprs.map(e => trimNonTopLevelAliases(e).asInstanceOf[NamedExpression])
+      Window(projectList, cleanedWindowExprs, partitionSpec.map(trimAliases),
+        orderSpec.map(trimAliases(_).asInstanceOf[SortOrder]), child)
+
+    case other =>
+      var stop = false
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Not for this release, but this makes me think that we are abusing aliases.  I would rather that resolved expressions past the analyzer move the names out of the subexpressions and into the `CreateStruct` expression itself.

**Comment by cloud-fan** [comment]
closing in favor of #8215
